src/Icicle/Common/Exp/Eval/Small.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Icicle.Common.Exp.Eval.Small
+
+where
+
+import qualified Data.Map as M
+
+import Icicle.Internal.Pretty
+import Icicle.Common.Base
+import Icicle.Common.Type
+import Icicle.Common.Value
+import Icicle.Common.Exp.Exp
+import Icicle.Common.Exp.Eval (EvalPrim)
+import qualified Icicle.Common.Exp.Eval as Big
+
+import P
+
+import Data.Either.Combinators (mapLeft)
+
+
+data RuntimeError n p
+  = RuntimeErrorPrimBadArgs p [Value n p]
+  | RuntimeErrorUnboundVar  (Name n)
+  | RuntimeErrorPrimBad p
+  | RuntimeErrorPrimEval (Big.RuntimeError n p)
+  deriving (Show, Eq)
+
+--------------------------------------------------------------------------------
+
+-- | Small step evaluation. Call-by-value for simplicity.
+--
+eval :: Eq n
+     => EvalPrim n p -> Exp n p
+     -> Either (RuntimeError n p) (Exp n p)
+eval evalPrim xx = case xx of
+  -- * Variables must have been substituted away.
+  XVar n
+    -> Left (RuntimeErrorUnboundVar n)
```

Comment:
I think this should just return "Right xx" - so we can partially evaluate open expressions